bench/bm_buffer.cpp

```diff
@@ -0,0 +1,153 @@
+    fair::meta::precondition(nProducer > 0);
+    fair::meta::precondition(nConsumer > 0);
+
+    constexpr int n_repeat = 8;
+
+    std::barrier barrier(nProducer + nConsumer + 1);
+
+    // init producers
+    std::atomic<int> threadID = 0;
+    std::vector<std::thread> producers;
+    for (int i = 0; i < nProducer; i++) {
+        producers.emplace_back([&]() {
+            // set thread affinity
+            setCpuAffinity(threadID++);
+            barrier.arrive_and_wait();
+            for (int rep = 0; rep < n_repeat; ++rep) {
+                BufferWriter auto writer = buffer.new_writer();
+                std::size_t nSamplesProduced = 0;
+                barrier.arrive_and_wait();
+                while (nSamplesProduced <= (min_samples + nProducer - 1) / nProducer) {
+                    writer.publish([](auto &) {}, vector_length);
+                    nSamplesProduced += vector_length;
+                }
+                barrier.arrive_and_wait();
+            }
+        });
+    }
+
+    // init consumers
+    std::vector<std::thread> consumers;
+    for (int i = 0; i < nConsumer; i++) {
+        consumers.emplace_back([&]() {
+            setCpuAffinity(threadID++);
+            barrier.arrive_and_wait();
+            for (int rep = 0; rep < n_repeat; ++rep) {
+                BufferReader auto reader = buffer.new_reader();
+                std::size_t nSamplesConsumed = 0;
+                barrier.arrive_and_wait();
+                while (nSamplesConsumed < min_samples) {
+                    if (reader.available() < vector_length) {
+                        continue;
+                    }
+                    const auto &input = reader.get(vector_length);
+                    nSamplesConsumed += input.size();
+
+                    if (!reader.consume(input.size())) {
+                        throw std::runtime_error(fmt::format("could not consume {} samples", input.size()));
+                    }
+                }
+                barrier.arrive_and_wait();
+            }
+        });
+    }
+
+    // all producers and consumer are ready, waiting to give the sign
+    barrier.arrive_and_wait();
+    ::benchmark::benchmark<n_repeat>(fmt::format("{}: {} writers -<{}>-> {} readers", name, nProducer, vector_length,
+                                                 nConsumer),
+                                     min_samples)
+            = [&]() {
+                  barrier.arrive_and_wait();
+                  barrier.arrive_and_wait();
+              };
+
+    // clean up
+    for (std::thread &thread : producers) thread.join();
+    for (std::thread &thread : consumers) thread.join();
+}
+
+inline const boost::ut::suite _buffer_tests = [] {
+    const uint64_t samples = 100'000'000; // minimum number of samples
+
+    for (int nP = 1; nP <= 4; nP *= 2) {
+        for (int nR = 1; nR <= 4; nR *= 2) {
+            if (not (nP == 1 and nR == 1)) {
+                benchmark::results::add_separator();
```

Comment:
I actually like the separator concept... we have to see that we get this benchmark upstream to the UT testing framework where the this separator is an reading aid in addition to the definition of suites... well done. :+1: 

![image](https://user-images.githubusercontent.com/46007894/224472042-bc1ef2a4-2f30-46c5-8961-6e46264398dc.png)